File: phantom-windows-app/Services/PhantomControlFrameParser.cs
using System;
using System.Collections.Generic;
using System.Linq;
using System.Text;
using System.Text.Json;
using SecureOverlay.Domain;

namespace SecureOverlay.Services
{
    public sealed class PhantomControlFrameParser
    {
        public const string ProtocolLine = "PHANTOM_CONTROL_V1";
        public const string BodyDelimiter = "\nPHANTOM_BODY\n";
        public const int MaxPrefixBytes = 32_768;
        private const int MaxRawPrefixBytes = 65_536;
        private static readonly string[] ThinkOpenTags = ["<thinking>", "<think>"];
        private static readonly string[] ThinkCloseTags = ["</thinking>", "</think>"];

        private static readonly HashSet<string> Actions = new(StringComparer.Ordinal) { "answer", "retrieve", "clarify" };
        private static readonly HashSet<string> QuestionTypes = new(StringComparer.Ordinal)
        {
            "behavioral", "technical", "coding", "system_design", "product_case", "motivation_fit",
            "personal_factual", "situational", "clarification", "unknown", "factual_lookup", "status_update",
            "decision_support", "objection_response", "risk_tradeoff", "brainstorm", "action_capture"
        };
        private static readonly HashSet<string> Intents = new(StringComparer.Ordinal) { "candidate_specific", "general", "hybrid", "ambiguous" };
        private static readonly HashSet<string> Bases = new(StringComparer.Ordinal) { "exact_evidence", "profile_synthesis", "universal_knowledge", "universal_synthesis", "clarification" };
        private static readonly HashSet<string> EntityTypes = new(StringComparer.Ordinal) { "none", "profile", "experience", "project", "document", "context_pack", "task" };

        private readonly HashSet<string> _allowedEntityIds;
        private readonly HashSet<string> _allowedDocumentIds;
        private readonly StringBuilder _prefix = new();
        private bool _bodyStarted;
        private bool _bodyHasContent;
        private int _receivedCharacters;

        public PhantomControlFrameParser(IEnumerable<string>? allowedEntityIds = null, IEnumerable<string>? allowedDocumentIds = null)
        {
            _allowedEntityIds = new HashSet<string>(allowedEntityIds ?? Array.Empty<string>(), StringComparer.Ordinal);
            _allowedDocumentIds = new HashSet<string>(allowedDocumentIds ?? Array.Empty<string>(), StringComparer.Ordinal);
        }

        public LiveTurnDecision? Decision { get; private set; }
        public int PrefixBytes { get; private set; }
        public bool HasReceivedChunks => _receivedCharacters > 0;

        public string Feed(string? chunk)
        {
            if (string.IsNullOrEmpty(chunk)) return string.Empty;
            _receivedCharacters += chunk.Length;
            if (_bodyStarted) return AcceptBody(chunk);

            _prefix.Append(chunk);
            var raw = _prefix.ToString();
            if (Encoding.UTF8.GetByteCount(raw) > MaxRawPrefixBytes)
            {
                throw new PhantomProtocolException("control_prefix_oversized");
            }

            var buffered = Normalize(raw);
            PrefixBytes = Encoding.UTF8.GetByteCount(buffered);
            if (TryCommitFrame(buffered, finalize: false, out var visible))
            {
                return visible;
            }

            if (PrefixBytes > MaxPrefixBytes)
            {
                throw new PhantomProtocolException("control_prefix_oversized");
            }

            return string.Empty;
        }

        public LiveTurnDecision Complete()
        {
            if (Decision is null)
            {
                var buffered = Normalize(_prefix.ToString());
                PrefixBytes = Encoding.UTF8.GetByteCount(buffered);
                _ = TryCommitFrame(buffered, finalize: true, out _);
            }

            var decision = Decision ?? throw new PhantomProtocolException("control_frame_incomplete");
            if (decision.Action != LiveCopilotAction.Retrieve && !_bodyHasContent)
                throw new PhantomProtocolException("answer_body_empty");
            return decision;
        }

        public static bool CanFallback(string code) => code is
            "control_frame_incomplete" or "control_prefix_invalid" or "control_prefix_oversized"
            or "control_json_invalid" or "answer_body_empty" or "control_enum_invalid"
            or "control_value_out_of_range" or "control_document_limit" or "control_document_duplicate"
            or "control_entity_unknown" or "control_document_unknown" or "retrieval_query_empty"
            or "unexpected_retrieval_query" or "clarification_basis_invalid" or "retrieve_body_not_empty";

        public static LiveTurnDecision FallbackAnswerDecision() => new(
            LiveCopilotAction.Answer,
            "unknown",
            "general",
            "universal_knowledge",
            "none",
            string.Empty,
            string.Empty,
            Array.Empty<string>(),
            40,
            false,
            0.5);

        public string GetFallbackAnswerText() => ExtractBareAnswer(_prefix.ToString());

        public static string ExtractAnswerBody(string response)
            => TryLocateFrame(Normalize(response), out _, out var rest, finalize: true) ? rest : string.Empty;

        public static string ResolveAnswer(string response)
        {
            var body = ExtractAnswerBody(response);
            return string.IsNullOrWhiteSpace(body) ? ExtractBareAnswer(response) : body;
        }

        public static string ExtractBareAnswer(string response)
        {
            var text = Normalize(response);
            if (TryLocateFrame(text, out _, out var rest, finalize: true))
            {
                return rest.Trim();
            }

            var protocolIndex = text.IndexOf(ProtocolLine, StringComparison.Ordinal);
            if (protocolIndex >= 0)
            {
                return string.Empty;
            }

            return text.Trim();
        }

        public static string Normalize(string text)
        {
            if (string.IsNullOrEmpty(text)) return string.Empty;
            return StripThink(text.Replace("\r\n", "\n").Replace('\r', '\n'));
        }

        private bool TryCommitFrame(string buffered, bool finalize, out string visible)
        {
            visible = string.Empty;
            if (_bodyStarted) return true;
            if (TryLocateFrame(buffered, out var json, out var rest, finalize))
            {
                if (string.IsNullOrWhiteSpace(json))
                {
                    throw new PhantomProtocolException("control_json_invalid");
                }

                Decision = ParseAndValidate(json);
                _bodyStarted = true;
                _prefix.Clear();
                visible = AcceptBody(rest);
                return true;
            }

            if (!LooksLikeBareAnswer(buffered, finalize))
            {
                return false;
            }

            Decision = FallbackAnswerDecision();
            _bodyStarted = true;
            _prefix.Clear();
            visible = AcceptBody(buffered);
            return true;
        }

        private static bool TryLocateFrame(string buffered, out string json, out string rest, bool finalize = false)
        {
            json = string.Empty;
            rest = string.Empty;
            var protocolIndex = buffered.IndexOf(ProtocolLine, StringComparison.Ordinal);
            var cursor = 0;
            if (protocolIndex >= 0)
            {
                cursor = protocolIndex + ProtocolLine.Length;
            }

            SkipControlNoise(buffered, ref cursor);
            if (cursor >= buffered.Length || buffered[cursor] != '{')
            {
                return false;
            }

            if (!TryReadJsonObject(buffered, cursor, out var jsonEnd))
            {
                return false;
            }

            json = buffered[cursor..jsonEnd].Trim();
            if (protocolIndex < 0 && json.IndexOf("\"action\"", StringComparison.Ordinal) < 0)
            {
                return false;
            }

            cursor = jsonEnd;
            SkipControlNoise(buffered, ref cursor);

            const string bodyToken = "PHANTOM_BODY";
            if (cursor < buffered.Length && StartsAt(buffered, cursor, bodyToken))
            {
                cursor += bodyToken.Length;
                if (cursor < buffered.Length && buffered[cursor] == '\n')
                {
                    cursor++;
                }

                rest = buffered[cursor..];
                return true;
            }

            var remaining = cursor < buffered.Length ? buffered[cursor..] : string.Empty;
            if (!finalize && (remaining.Length == 0 || IsTokenPrefix(remaining, bodyToken)))
            {
                return false;
            }

            rest = remaining;
            return true;
        }

        private static bool LooksLikeBareAnswer(string buffered, bool finalize)
        {
            var text = buffered.TrimStart();
            if (text.Length == 0) return false;
            if (text.Contains(ProtocolLine, StringComparison.Ordinal)) return false;
            if (IsTokenPrefix(text, ProtocolLine)) return false;
            if (text[0] == '{') return false;
            if (text.StartsWith("```", StringComparison.Ordinal))
            {
                return finalize && text.Length >= 80;
            }

            return finalize || text.Length >= 20;
        }

        private static void SkipControlNoise(string text, ref int index)
        {
            while (index < text.Length)
            {
                SkipWhitespace(text, ref index);
                if (!StartsAt(text, index, "```"))
                {
                    break;
                }

                index += 3;
                while (index < text.Length && text[index] != '\n' && text[index] != '{')
                {
                    index++;
                }

                if (index < text.Length && text[index] == '\n')
                {
                    index++;
                }
            }
        }

        private static void SkipWhitespace(string text, ref int index)
        {
            while (index < text.Length && char.IsWhiteSpace(text[index]))
            {
                index++;
            }
        }

        private static bool StartsAt(string text, int index, string token)
        {
            return index + token.Length <= text.Length
                && string.CompareOrdinal(text, index, token, 0, token.Length) == 0;
        }

        private static bool IsTokenPrefix(string remaining, string token)
            => remaining.Length > 0 && remaining.Length < token.Length
                && token.StartsWith(remaining, StringComparison.Ordinal);

        private static bool TryReadJsonObject(string text, int start, out int endExclusive)
        {
            endExclusive = start;
            if (start >= text.Length || text[start] != '{') return false;

            var depth = 0;
            var inString = false;
            var escape = false;
            for (var i = start; i < text.Length; i++)
            {
                var c = text[i];
                if (inString)
                {
                    if (escape)
                    {
                        escape = false;
                        continue;
                    }

                    if (c == '\\')
                    {
                        escape = true;
                        continue;
                    }

                    if (c == '"') inString = false;
                    continue;
                }

                if (c == '"')
                {
                    inString = true;
                    continue;
                }

                if (c == '{') depth++;
                else if (c == '}')
                {
                    depth--;
                    if (depth == 0)
                    {
                        endExclusive = i + 1;
                        return true;
                    }
                }
            }

            return false;
        }

        private static string StripThink(string text)
        {
            var value = text;
            while (true)
            {
                var openIndex = IndexOfToken(value, ThinkOpenTags, 0, out var openToken);
                if (openIndex < 0) return value;

                var afterOpen = openIndex + openToken.Length;
                var closeIndex = IndexOfToken(value, ThinkCloseTags, afterOpen, out var closeToken);
                if (closeIndex >= 0)
                {
                    value = value.Remove(openIndex, closeIndex + closeToken.Length - openIndex);
                    continue;
                }

                var protocolIndex = value.IndexOf(ProtocolLine, afterOpen, StringComparison.Ordinal);
                if (protocolIndex >= 0)
                {
                    value = value.Remove(openIndex, protocolIndex - openIndex);
                    continue;
                }

                return value[..openIndex];
            }
        }

        private static int IndexOfToken(string text, string[] tokens, int start, out string token)
        {
            token = string.Empty;
            var best = -1;
            foreach (var candidate in tokens)
            {
                var index = text.IndexOf(candidate, start, StringComparison.OrdinalIgnoreCase);
                if (index < 0) continue;
                if (best < 0 || index < best || (index == best && candidate.Length > token.Length))
                {
                    best = index;
                    token = candidate;
                }
            }

            return best;
        }

        private string AcceptBody(string body)
        {
            if (!string.IsNullOrWhiteSpace(body)) _bodyHasContent = true;
            return Decision?.Action == LiveCopilotAction.Retrieve ? string.Empty : body;
        }

        private LiveTurnDecision ParseAndValidate(string json)
        {
            JsonElement root;
            try
            {
                using var document = JsonDocument.Parse(json);
                if (document.RootElement.ValueKind != JsonValueKind.Object)
                    throw new PhantomProtocolException("control_json_invalid");
                root = document.RootElement.Clone();
            }
            catch (JsonException)
            {
                throw new PhantomProtocolException("control_json_invalid");
            }

            var actionRaw = ReadString(root, "action");
            if (!Actions.Contains(actionRaw))
                throw new PhantomProtocolException("control_enum_invalid");

            var questionType = CoerceEnum(ReadString(root, "questionType"), QuestionTypes, "unknown");
            var intent = CoerceEnum(ReadString(root, "intent"), Intents, "general");
            var answerBasis = CoerceEnum(ReadString(root, "answerBasis"), Bases, "universal_knowledge");
            var entityType = CoerceEnum(ReadString(root, "entityType"), EntityTypes, "none");
            var entityId = TrimBound(ReadString(root, "entityId"), 160);
            if (!string.IsNullOrEmpty(entityId) && !_allowedEntityIds.Contains(entityId))
            {
                entityId = string.Empty;
                entityType = "none";
            }

            var retrievalQuery = TrimBound(ReadString(root, "retrievalQuery"), 500);
            var preferredDocumentIds = ReadStringArray(root, "preferredDocumentIds")
                .Where(id => !string.IsNullOrWhiteSpace(id) && id.Length <= 160)
                .Distinct(StringComparer.Ordinal)
                .Where(_allowedDocumentIds.Contains)
                .Take(8)
                .ToArray();

            var action = actionRaw switch
            {
                "retrieve" => LiveCopilotAction.Retrieve,
                "clarify" => LiveCopilotAction.Clarify,
                _ => LiveCopilotAction.Answer
            };
            if (action != LiveCopilotAction.Retrieve)
                retrievalQuery = string.Empty;
            if (action == LiveCopilotAction.Clarify)
                answerBasis = "clarification";

            var confidence = ReadDouble(root, "confidence", 0.5);
            if (double.IsNaN(confidence) || double.IsInfinity(confidence))
                confidence = 0.5;
            confidence = Math.Clamp(confidence, 0, 1);

            return new LiveTurnDecision(
                action,
                questionType,
                intent,
                answerBasis,
                entityType,
                entityId,
                retrievalQuery,
                preferredDocumentIds,
                ClampTargetSeconds(questionType, ReadInt(root, "targetSeconds", 0)),
                ReadBool(root, "allowCode"),
                confidence);
        }

        private static string CoerceEnum(string value, HashSet<string> allowed, string fallback)
            => allowed.Contains(value) ? value : fallback;

        private static string TrimBound(string value, int max)
            => value.Length <= max ? value : value[..max];

        private static JsonElement? FindProperty(JsonElement root, string name)
        {
            foreach (var property in root.EnumerateObject())
            {
                if (string.Equals(property.Name, name, StringComparison.OrdinalIgnoreCase))
                    return property.Value;
            }

            return null;
        }

        private static string ReadString(JsonElement root, string name)
        {
            var property = FindProperty(root, name);
            if (property is null) return string.Empty;
            return property.Value.ValueKind switch
            {
                JsonValueKind.String => property.Value.GetString() ?? string.Empty,
                JsonValueKind.Number => property.Value.GetRawText(),
                JsonValueKind.True => "true",
                JsonValueKind.False => "false",
                JsonValueKind.Null => string.Empty,
                _ => string.Empty
            };
        }

        private static int ReadInt(JsonElement root, string name, int fallback)
        {
            var property = FindProperty(root, name);
            if (property is null) return fallback;
            if (property.Value.ValueKind == JsonValueKind.Number && property.Value.TryGetInt32(out var number))
                return number;
            return int.TryParse(ReadString(root, name), out var parsed) ? parsed : fallback;
        }

        private static double ReadDouble(JsonElement root, string name, double fallback)
        {
            var property = FindProperty(root, name);
            if (property is null) return fallback;
            if (property.Value.ValueKind == JsonValueKind.Number && property.Value.TryGetDouble(out var number))
                return number;
            return double.TryParse(ReadString(root, name), out var parsed) ? parsed : fallback;
        }

        private static bool ReadBool(JsonElement root, string name)
        {
            var property = FindProperty(root, name);
            if (property is null) return false;
            if (property.Value.ValueKind == JsonValueKind.True) return true;
            if (property.Value.ValueKind == JsonValueKind.False) return false;
            return bool.TryParse(ReadString(root, name), out var parsed) && parsed;
        }

        private static List<string> ReadStringArray(JsonElement root, string name)
        {
            var property = FindProperty(root, name);
            if (property is null || property.Value.ValueKind != JsonValueKind.Array)
                return new List<string>();
            var values = new List<string>();
            foreach (var item in property.Value.EnumerateArray())
            {
                if (item.ValueKind == JsonValueKind.String)
                    values.Add(item.GetString() ?? string.Empty);
            }

            return values;
        }

        private static int ClampTargetSeconds(string questionType, int value)
        {
            var (minimum, maximum) = questionType switch
            {
                "behavioral" => (25, 45),
                "technical" => (20, 40),
                "coding" => (30, 90),
                "system_design" => (40, 75),
                "product_case" => (30, 60),
                "motivation_fit" => (20, 40),
                "personal_factual" => (10, 30),
                "situational" => (20, 40),
                "clarification" or "unknown" => (5, 20),
                "factual_lookup" or "status_update" or "decision_support" or "objection_response"
                    or "risk_tradeoff" or "brainstorm" or "action_capture" => (10, 40),
                _ => (15, 45)
            };
            return Math.Clamp(value, minimum, maximum);
        }

    }
}

File: phantom-mac-app/Sources/Phantom/Services/PhantomControlFrameParser.swift
import Foundation

enum CopilotMode: String, Codable, CaseIterable, Hashable {
    case interview
    case briefing
}

enum InterviewDeliveryStyle: String, Codable, CaseIterable, Hashable {
    case standard
    case desi
}

enum LiveCopilotAction: String, Codable {
    case answer
    case retrieve
    case clarify
}

struct LiveTurnDecision: Equatable {
    let action: LiveCopilotAction
    let questionType: String
    let intent: String
    let answerBasis: String
    let entityType: String
    let entityId: String
    let retrievalQuery: String
    let preferredDocumentIds: [String]
    let targetSeconds: Int
    let allowCode: Bool
    let confidence: Double
    let protocolVersion = 1

    func withRetrieval(query: String, preferredDocumentIds: [String]) -> LiveTurnDecision {
        LiveTurnDecision(
            action: .retrieve,
            questionType: questionType,
            intent: intent,
            answerBasis: answerBasis,
            entityType: entityType,
            entityId: entityId,
            retrievalQuery: query,
            preferredDocumentIds: preferredDocumentIds,
            targetSeconds: targetSeconds,
            allowCode: allowCode,
            confidence: confidence
        )
    }
}

struct LiveCopilotRetrieval {
    let status: String
    let snippets: [KnowledgeSnippet]
    let kbRevision: String
}

struct LiveCopilotResult {
    let answer: String
    let decision: LiveTurnDecision
    let modelCallCount: Int
    let protocolRetryCount: Int
    let retrievalStatus: String
    let activeEvidence: [KnowledgeSnippet]
}

struct PhantomProtocolError: LocalizedError {
    let code: String
    var errorDescription: String? { "The selected model returned an invalid live-response header." }
}

final class PhantomControlFrameParser {
    static let protocolLine = "PHANTOM_CONTROL_V1"
    static let bodyDelimiter = "\nPHANTOM_BODY\n"
    static let maxPrefixBytes = 32_768
    private static let maxRawPrefixBytes = 65_536
    private static let thinkOpenTags = ["<thinking>", "<think>"]
    private static let thinkCloseTags = ["</thinking>", "</think>"]

    private static let actions = Set(["answer", "retrieve", "clarify"])
    private static let questionTypes = Set([
        "behavioral", "technical", "coding", "system_design", "product_case", "motivation_fit",
        "personal_factual", "situational", "clarification", "unknown", "factual_lookup", "status_update",
        "decision_support", "objection_response", "risk_tradeoff", "brainstorm", "action_capture"
    ])
    private static let intents = Set(["candidate_specific", "general", "hybrid", "ambiguous"])
    private static let bases = Set(["exact_evidence", "profile_synthesis", "universal_knowledge", "universal_synthesis", "clarification"])
    private static let entityTypes = Set(["none", "profile", "experience", "project", "document", "context_pack", "task"])

    private let allowedEntityIds: Set<String>
    private let allowedDocumentIds: Set<String>
    private var prefix = ""
    private var bodyStarted = false
    private var bodyHasContent = false
    private(set) var decision: LiveTurnDecision?
    private(set) var prefixBytes = 0
    private(set) var hasReceivedChunks = false

    init(allowedEntityIds: [String] = [], allowedDocumentIds: [String] = []) {
        self.allowedEntityIds = Set(allowedEntityIds)
        self.allowedDocumentIds = Set(allowedDocumentIds)
    }

    func feed(_ chunk: String) throws -> String {
        guard !chunk.isEmpty else { return "" }
        hasReceivedChunks = true
        if bodyStarted { return acceptBody(chunk) }
        prefix += chunk
        let rawBytes = prefix.lengthOfBytes(using: .utf8)
        guard rawBytes <= Self.maxRawPrefixBytes else { throw PhantomProtocolError(code: "control_prefix_oversized") }
        let buffered = Self.normalize(prefix)
        prefixBytes = buffered.lengthOfBytes(using: .utf8)
        if let visible = try commitFrame(buffered, finalize: false) {
            return visible
        }
        guard prefixBytes <= Self.maxPrefixBytes else { throw PhantomProtocolError(code: "control_prefix_oversized") }
        return ""
    }

    func complete() throws -> LiveTurnDecision {
        if decision == nil {
            let buffered = Self.normalize(prefix)
            prefixBytes = buffered.lengthOfBytes(using: .utf8)
            _ = try commitFrame(buffered, finalize: true)
        }
        guard let decision else { throw PhantomProtocolError(code: "control_frame_incomplete") }
        if decision.action != .retrieve, !bodyHasContent { throw PhantomProtocolError(code: "answer_body_empty") }
        return decision
    }

    static func canFallback(_ code: String) -> Bool {
        [
            "control_frame_incomplete", "control_prefix_invalid", "control_prefix_oversized",
            "control_json_invalid", "answer_body_empty", "control_enum_invalid",
            "control_value_out_of_range", "control_document_limit", "control_document_duplicate",
            "control_entity_unknown", "control_document_unknown", "retrieval_query_empty",
            "unexpected_retrieval_query", "clarification_basis_invalid", "retrieve_body_not_empty"
        ].contains(code)
    }

    static func fallbackAnswerDecision() -> LiveTurnDecision {
        LiveTurnDecision(
            action: .answer, questionType: "unknown", intent: "general",
            answerBasis: "universal_knowledge", entityType: "none", entityId: "",
            retrievalQuery: "", preferredDocumentIds: [], targetSeconds: 40,
            allowCode: false, confidence: 0.5
        )
    }

    func fallbackAnswerText() -> String { Self.extractBareAnswer(prefix) }

    static func extractAnswerBody(_ response: String) -> String {
        let text = normalize(response)
        return locateFrame(text, finalize: true)?.rest ?? ""
    }

    static func resolveAnswer(_ response: String) -> String {
        let body = extractAnswerBody(response)
        if !body.trimmingCharacters(in: .whitespacesAndNewlines).isEmpty { return body }
        return extractBareAnswer(response)
    }

    static func extractBareAnswer(_ response: String) -> String {
        let text = normalize(response)
        if let frame = locateFrame(text, finalize: true) { return frame.rest.trimmingCharacters(in: .whitespacesAndNewlines) }
        if text.contains(protocolLine) { return "" }
        return text.trimmingCharacters(in: .whitespacesAndNewlines)
    }

    static func normalize(_ text: String) -> String {
        guard !text.isEmpty else { return "" }
        return stripThink(text.replacingOccurrences(of: "\r\n", with: "\n").replacingOccurrences(of: "\r", with: "\n"))
    }

    private func commitFrame(_ buffered: String, finalize: Bool) throws -> String? {
        if bodyStarted { return "" }
        if let frame = Self.locateFrame(buffered, finalize: finalize) {
            guard !frame.json.trimmingCharacters(in: .whitespacesAndNewlines).isEmpty else {
                throw PhantomProtocolError(code: "control_json_invalid")
            }
            decision = try Self.parse(frame.json, allowedEntityIds: allowedEntityIds, allowedDocumentIds: allowedDocumentIds)
            bodyStarted = true
            prefix = ""
            return acceptBody(frame.rest)
        }
        guard Self.looksLikeBareAnswer(buffered, finalize: finalize) else { return nil }
        decision = Self.fallbackAnswerDecision()
        bodyStarted = true
        prefix = ""
        return acceptBody(buffered)
    }

    private static func locateFrame(_ buffered: String, finalize: Bool = false) -> (json: String, rest: String)? {
        let protocolRange = buffered.range(of: protocolLine)
        var cursor = buffered.startIndex
        if let protocolRange {
            cursor = protocolRange.upperBound
        }
        skipControlNoise(buffered, from: &cursor)
        guard cursor < buffered.endIndex, buffered[cursor] == "{" else { return nil }
        guard let jsonEnd = readJsonObjectEnd(buffered, from: cursor) else { return nil }
        let json = String(buffered[cursor..<jsonEnd]).trimmingCharacters(in: .whitespacesAndNewlines)
        if protocolRange == nil, !json.contains("\"action\"") {
            return nil
        }
        cursor = jsonEnd
        skipControlNoise(buffered, from: &cursor)

        let bodyToken = "PHANTOM_BODY"
        if cursor < buffered.endIndex, buffered[cursor...].hasPrefix(bodyToken) {
            cursor = buffered.index(cursor, offsetBy: bodyToken.count)
            if cursor < buffered.endIndex, buffered[cursor] == "\n" {
                cursor = buffered.index(after: cursor)
            }
            return (json, String(buffered[cursor...]))
        }

        let remaining = cursor < buffered.endIndex ? String(buffered[cursor...]) : ""
        if !finalize && (remaining.isEmpty || isTokenPrefix(remaining, of: bodyToken)) {
            return nil
        }
        return (json, remaining)
    }

    private static func looksLikeBareAnswer(_ buffered: String, finalize: Bool) -> Bool {
        let text = buffered.trimmingCharacters(in: .whitespacesAndNewlines)
        guard !text.isEmpty else { return false }
        if text.contains(protocolLine) { return false }
        let leading = buffered.drop(while: { $0.isWhitespace })
        let leadingText = String(leading)
        if isTokenPrefix(leadingText, of: protocolLine) { return false }
        guard let first = leadingText.first else { return false }
        if first == "{" { return false }
        if leadingText.hasPrefix("```") { return finalize && leadingText.count >= 80 }
        return finalize || leadingText.count >= 20
    }

    private static func skipControlNoise(_ text: String, from index: inout String.Index) {
        while index < text.endIndex {
            skipWhitespace(text, from: &index)
            guard index < text.endIndex, text[index...].hasPrefix("```") else { break }
            index = text.index(index, offsetBy: 3)
            while index < text.endIndex, text[index] != "\n", text[index] != "{" {
                index = text.index(after: index)
            }
            if index < text.endIndex, text[index] == "\n" {
                index = text.index(after: index)
            }
        }
    }

    private static func skipWhitespace(_ text: String, from index: inout String.Index) {
        while index < text.endIndex, text[index].isWhitespace {
            index = text.index(after: index)
        }
    }

    private static func isTokenPrefix(_ remaining: String, of token: String) -> Bool {
        !remaining.isEmpty && remaining.count < token.count && token.hasPrefix(remaining)
    }

    private static func readJsonObjectEnd(_ text: String, from start: String.Index) -> String.Index? {
        guard start < text.endIndex, text[start] == "{" else { return nil }
        var depth = 0
        var inString = false
        var escape = false
        var index = start
        while index < text.endIndex {
            let character = text[index]
            let next = text.index(after: index)
            if inString {
                if escape {
                    escape = false
                } else if character == "\\" {
                    escape = true
                } else if character == "\"" {
                    inString = false
                }
                index = next
                continue
            }
            if character == "\"" {
                inString = true
            } else if character == "{" {
                depth += 1
            } else if character == "}" {
                depth -= 1
                if depth == 0 {
                    return next
                }
            }
            index = next
        }
        return nil
    }

    private static func stripThink(_ text: String) -> String {
        var value = text
        while true {
            guard let open = firstToken(thinkOpenTags, in: value, from: value.startIndex) else { return value }
            let afterOpen = value.index(open.range.lowerBound, offsetBy: open.token.count)
            if let close = firstToken(thinkCloseTags, in: value, from: afterOpen) {
                value.removeSubrange(open.range.lowerBound..<close.range.upperBound)
                continue
            }
            if let protocolRange = value.range(of: protocolLine, range: afterOpen..<value.endIndex) {
                value.removeSubrange(open.range.lowerBound..<protocolRange.lowerBound)
                continue
            }
            return String(value[..<open.range.lowerBound])
        }
    }

    private static func firstToken(_ tokens: [String], in text: String, from start: String.Index) -> (token: String, range: Range<String.Index>)? {
        var best: (token: String, range: Range<String.Index>)?
        for token in tokens {
            guard let range = text.range(of: token, options: .caseInsensitive, range: start..<text.endIndex) else { continue }
            if best == nil || range.lowerBound < best!.range.lowerBound || (range.lowerBound == best!.range.lowerBound && token.count > best!.token.count) {
                best = (token, range)
            }
        }
        return best
    }

    private func acceptBody(_ body: String) -> String {
        if !body.trimmingCharacters(in: .whitespacesAndNewlines).isEmpty { bodyHasContent = true }
        return decision?.action == .retrieve ? "" : body
    }

    private static func parse(_ json: String, allowedEntityIds: Set<String>, allowedDocumentIds: Set<String>) throws -> LiveTurnDecision {
        let data = Data(json.utf8)
        guard let object = (try? JSONSerialization.jsonObject(with: data)) as? [String: Any] else {
            throw PhantomProtocolError(code: "control_json_invalid")
        }
        let actionRaw = stringValue(object, "action")
        guard actions.contains(actionRaw), let action = LiveCopilotAction(rawValue: actionRaw) else {
            throw PhantomProtocolError(code: "control_enum_invalid")
        }
        let questionType = coerce(stringValue(object, "questionType"), allowed: questionTypes, fallback: "unknown")
        let intent = coerce(stringValue(object, "intent"), allowed: intents, fallback: "general")
        var answerBasis = coerce(stringValue(object, "answerBasis"), allowed: bases, fallback: "universal_knowledge")
        var entityType = coerce(stringValue(object, "entityType"), allowed: entityTypes, fallback: "none")
        var entityId = String(stringValue(object, "entityId").prefix(160))
        if !entityId.isEmpty, !allowedEntityIds.contains(entityId) {
            entityId = ""
            entityType = "none"
        }
        var retrievalQuery = String(stringValue(object, "retrievalQuery").prefix(500))
        let preferredDocumentIds = stringArray(object, "preferredDocumentIds")
            .filter { !$0.isEmpty && $0.count <= 160 }
            .reduce(into: [String]()) { result, id in
                if !result.contains(id) { result.append(id) }
            }
            .filter(allowedDocumentIds.contains)
            .prefix(8)
            .map { $0 }
        if action != .retrieve { retrievalQuery = "" }
        if action == .clarify { answerBasis = "clarification" }
        var confidence = doubleValue(object, "confidence", 0.5)
        if confidence.isNaN || confidence.isInfinite { confidence = 0.5 }
        confidence = min(max(confidence, 0), 1)
        return LiveTurnDecision(
            action: action, questionType: questionType, intent: intent, answerBasis: answerBasis,
            entityType: entityType, entityId: entityId,
            retrievalQuery: retrievalQuery.trimmingCharacters(in: .whitespacesAndNewlines),
            preferredDocumentIds: preferredDocumentIds,
            targetSeconds: clampTargetSeconds(intValue(object, "targetSeconds"), questionType: questionType),
            allowCode: boolValue(object, "allowCode"), confidence: confidence
        )
    }

    private static func lookup(_ object: [String: Any], _ key: String) -> Any? {
        if let value = object[key] { return value }
        let lowered = key.lowercased()
        return object.first { $0.key.lowercased() == lowered }?.value
    }

    private static func coerce(_ value: String, allowed: Set<String>, fallback: String) -> String {
        allowed.contains(value) ? value : fallback
    }

    private static func stringValue(_ object: [String: Any], _ key: String) -> String {
        switch lookup(object, key) {
        case let value as String: return value
        case let value as NSNumber: return value.stringValue
        default: return ""
        }
    }

    private static func intValue(_ object: [String: Any], _ key: String) -> Int {
        switch lookup(object, key) {
        case let value as Int: return value
        case let value as NSNumber: return value.intValue
        case let value as String: return Int(value) ?? 0
        default: return 0
        }
    }

    private static func doubleValue(_ object: [String: Any], _ key: String, _ fallback: Double) -> Double {
        switch lookup(object, key) {
        case let value as Double: return value
        case let value as NSNumber: return value.doubleValue
        case let value as String: return Double(value) ?? fallback
        default: return fallback
        }
    }

    private static func boolValue(_ object: [String: Any], _ key: String) -> Bool {
        switch lookup(object, key) {
        case let value as Bool: return value
        case let value as NSNumber: return value.boolValue
        case let value as String: return (value as NSString).boolValue
        default: return false
        }
    }

    private static func stringArray(_ object: [String: Any], _ key: String) -> [String] {
        (lookup(object, key) as? [Any] ?? []).compactMap { item in
            item as? String
        }
    }

    private static func clampTargetSeconds(_ value: Int, questionType: String) -> Int {
        let bounds: ClosedRange<Int>
        switch questionType {
        case "behavioral": bounds = 25...45
        case "technical": bounds = 20...40
        case "coding": bounds = 30...90
        case "system_design": bounds = 40...75
        case "product_case": bounds = 30...60
        case "motivation_fit": bounds = 20...40
        case "personal_factual": bounds = 10...30
        case "situational": bounds = 20...40
        case "clarification", "unknown": bounds = 5...20
        case "factual_lookup", "status_update", "decision_support", "objection_response",
             "risk_tradeoff", "brainstorm", "action_capture": bounds = 10...40
        default: bounds = 15...45
        }
        return min(max(value, bounds.lowerBound), bounds.upperBound)
    }

}
